Refactor incorrect `CoreRerouteTest` and `RouteAlternativesTest`

An issue caught by an Instrumentation test found while working on https://github.com/mapbox/mapbox-navigation-android/pull/4419 https://app.circleci.com/pipelines/github/mapbox/mapbox-navigation-android/12041/workflows/6cc8f997-07d1-49f4-a898-9243e7ace970/jobs/48760 https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.54eb2dd57e959221/matrices/5010895745904766073/executions/bs.1e3c7202ff856a89/testcases/2/errors

```
java.lang.AssertionError
at org.junit.Assert.fail(Assert.java:87)
at org.junit.Assert.assertTrue(Assert.java:42)
at org.junit.Assert.assertTrue(Assert.java:53)
at com.mapbox.navigation.instrumentation_tests.core.RouteAlternativesTest.expect_faster_route_alternatives(RouteAlternativesTest.kt:79)
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:134)
at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
at com.schibsted.spain.barista.rule.cleardata.ClearDatabaseRule$apply$1.evaluate(ClearDatabaseRule.kt:30)
at com.schibsted.spain.barista.rule.cleardata.ClearPreferencesRule$1.evaluate(ClearPreferencesRule.java:26)
at com.schibsted.spain.barista.rule.cleardata.ClearFilesRule$apply$1.evaluate(ClearFilesRule.kt:25)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
at org.junit.rules.RunRules.evaluate(RunRules.java:20)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runners.Suite.runChild(Suite.java:128)
at org.junit.runners.Suite.runChild(Suite.java:27)
at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2074)
```

cc @RingerJK @kmadsen

Last run `AlternativeRouteSelectionTest#selectAlternateRoute` failed https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.54eb2dd57e959221/matrices/4908335199249285018/executions/bs.5c115477058a2782/testcases/6/logs

```
06-07 02:15:38.996: W/System.err(21218): java.lang.AssertionError: expected:<DirectionsRoute{routeIndex=0, distance=1683.8, duration=207.3, durationTypical=null, geometry=mvl`gAnalshFa@iVUuMG}D[{Rq@y`@ScLaB{aA{@gi@[kQGmEIqE[yQ{@uh@aAgl@_@{Ta@qU_Agk@wl@hBiRj@eELag@|AiDJkELimB~FusBlGwAg|@IuDGwDoAwu@GmEsDLcIVooAxDqCHoFPaGP}FPTlN, weight=365.4, weightName=routability, legs=[RouteLeg{distance=1683.8, duration=207.3, durationTypical=null, summary=Fulton Street, 27th Avenue, admins=null, steps=[LegStep{distance=628.0, duration=63.099999999999994, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=mvl`gAnalshFa@iVUuMG}D[{Rq@y`@ScLaB{aA{@gi@[kQGmEIqE[yQ{@uh@aAgl@_@{Ta@qU_Agk@, name=Fulton Street, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.49348, 37.772151], bearingBefore=0.0, bearingAfter=87.0, instruction=Head east on Fulton Street, type=depart, modifier=null, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=628.0, announcement=Head east on Fulton Street, then turn left onto 27th Avenue, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Head east on Fulton Street, then turn left onto <say-as interpret-as="address">27th</say-as> Avenue</prosody></amazon:effect></speak>}, VoiceInstructions{distanceAlongGeometry=149.3, announcement=Turn left onto 27th Avenue, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Turn left onto <say-as interpret-as="address">27th</say-as> Avenue</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=628.0, primary=BannerText{text=27th Avenue, components=[BannerComponents{text=27th Avenue, type=text, subType=null, abbreviation=27th Ave, abbreviationPriority=0, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=turn, modifier=left, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=107.7, intersections=[StepIntersection{rawLocation=[-122.49348, 37.772151], bearings=[87], classes=null, entry=[true], in=null, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.492777, 37.772183], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.491708, 37.772232], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.490638, 37.772281], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.489565, 37.772329], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, true, false], in=3, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.488492, 37.772378], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.487418, 37.772427], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=623.9, duration=72.3, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=wjm`gAxc~rhFwl@hBiRj@eELag@|AiDJk
06-07 02:15:38.997: W/System.err(21218): 	at org.junit.Assert.fail(Assert.java:89)
06-07 02:15:38.997: W/System.err(21218): 	at org.junit.Assert.failNotEquals(Assert.java:835)
06-07 02:15:38.997: W/System.err(21218): 	at org.junit.Assert.assertEquals(Assert.java:120)
06-07 02:15:38.997: W/System.err(21218): 	at org.junit.Assert.assertEquals(Assert.java:146)
06-07 02:15:38.997: W/System.err(21218): 	at com.mapbox.navigation.instrumentation_tests.ui.routeline.AlternativeRouteSelectionTest$selectAlternateRoute$1$2.onRouteProgressChanged(AlternativeRouteSelectionTest.kt:127)
06-07 02:15:38.997: W/System.err(21218): 	at com.mapbox.navigation.core.trip.session.MapboxTripSession.updateRouteProgress(MapboxTripSession.kt:557)
06-07 02:15:38.997: W/System.err(21218): 	at com.mapbox.navigation.core.trip.session.MapboxTripSession.access$updateRouteProgress(MapboxTripSession.kt:55)
06-07 02:15:38.997: W/System.err(21218): 	at com.mapbox.navigation.core.trip.session.MapboxTripSession$navigatorObserver$1.onStatus(MapboxTripSession.kt:217)
06-07 02:15:38.997: W/System.err(21218): 	at android.os.MessageQueue.nativePollOnce(Native Method)
06-07 02:15:38.997: W/System.err(21218): 	at android.os.MessageQueue.next(MessageQueue.java:326)
06-07 02:15:38.997: W/System.err(21218): 	at java.lang.reflect.Method.invoke(Native Method)
06-07 02:15:38.997: W/System.err(21218): 	at androidx.test.espresso.base.Interrogator.getNextMessage(Interrogator.java:199)
06-07 02:15:38.997: W/System.err(21218): 	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:140)
06-07 02:15:38.997: W/System.err(21218): 	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
06-07 02:15:38.997: W/System.err(21218): 	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
06-07 02:15:38.997: W/System.err(21218): 	at androidx.test.espresso.Espresso$1.run(Espresso.java:277)
06-07 02:15:38.997: W/System.err(21218): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458)
06-07 02:15:38.997: W/System.err(21218): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
06-07 02:15:38.997: W/System.err(21218): 	at android.os.Handler.handleCallback(Handler.java:873)
06-07 02:15:38.998: W/System.err(21218): 	at android.os.Handler.dispatchMessage(Handler.java:99)
06-07 02:15:38.998: W/System.err(21218): 	at android.os.Looper.loop(Looper.java:193)
06-07 02:15:38.998: W/System.err(21218): 	at android.app.ActivityThread.main(ActivityThread.java:6718)
06-07 02:15:38.998: W/System.err(21218): 	at java.lang.reflect.Method.invoke(Native Method)
06-07 02:15:38.998: W/System.err(21218): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
06-07 02:15:38.998: W/System.err(21218): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
06-07 02:15:39.000: E/libc++abi(21218): terminating with uncaught exception of type jni::PendingJavaException
```

Noting that It's not failing locally.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -357,6 +357,7 @@ commands:
               --device model=star2lte,version=28,locale=de,orientation=landscape \
               --device model=flame,version=30,locale=en,orientation=portrait \
               --use-orchestrator \
+              --directories-to-pull=/sdcard/Download/mapbox_test \
               --timeout 15m
 
   run-firebase-robo:

--- a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt
+++ b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt
@@ -4,9 +4,12 @@ import androidx.test.espresso.Espresso
 import com.mapbox.api.directions.v5.DirectionsCriteria
 import com.mapbox.api.directions.v5.models.DirectionsRoute
 import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
 import com.mapbox.geojson.Point
 import com.mapbox.navigation.base.extensions.applyDefaultNavigationOptions
 import com.mapbox.navigation.base.extensions.applyLanguageAndVoiceUnitOptions
+import com.mapbox.navigation.base.options.HistoryRecorderOptions
 import com.mapbox.navigation.base.options.NavigationOptions
 import com.mapbox.navigation.base.route.RouterCallback
 import com.mapbox.navigation.base.route.RouterFailure
@@ -18,6 +21,7 @@ import com.mapbox.navigation.instrumentation_tests.R
 import com.mapbox.navigation.instrumentation_tests.activity.EmptyTestActivity
 import com.mapbox.navigation.instrumentation_tests.utils.MapboxNavigationRule
 import com.mapbox.navigation.instrumentation_tests.utils.assertions.RouteProgressStateTransitionAssertion
+import com.mapbox.navigation.instrumentation_tests.utils.history.MapboxHistoryTestRule
 import com.mapbox.navigation.instrumentation_tests.utils.http.MockDirectionsRequestHandler
 import com.mapbox.navigation.instrumentation_tests.utils.idling.RouteProgressStateIdlingResource
 import com.mapbox.navigation.instrumentation_tests.utils.location.MockLocationReplayerRule
@@ -26,11 +30,12 @@ import com.mapbox.navigation.instrumentation_tests.utils.routes.MockRoutesProvid
 import com.mapbox.navigation.instrumentation_tests.utils.runOnMainSync
 import com.mapbox.navigation.testing.ui.BaseTest
 import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import com.mapbox.navigation.utils.internal.logE
 import org.junit.Before
 import org.junit.Rule
 import org.junit.Test
+import java.util.concurrent.CountDownLatch
 
-// TODO: Refactor https://github.com/mapbox/mapbox-navigation-android/issues/4429
 class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
 
     @get:Rule
@@ -38,6 +43,9 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
 
     @get:Rule
     val mockLocationReplayerRule = MockLocationReplayerRule(mockLocationUpdatesRule)
+
+    @get:Rule
+    val mapboxHistoryTestRule = MapboxHistoryTestRule()
 
     private lateinit var mapboxNavigation: MapboxNavigation
 
@@ -48,11 +56,18 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
     fun setup() {
         Espresso.onIdle()
 
-        mapboxNavigation = MapboxNavigationProvider.create(
-            NavigationOptions.Builder(activity)
-                .accessToken(getMapboxAccessTokenFromResources(activity))
-                .build()
-        )
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .historyRecorderOptions(
+                        HistoryRecorderOptions.Builder()
+                            .build()
+                    )
+                    .build()
+            )
+            mapboxHistoryTestRule.historyRecorder = mapboxNavigation.historyRecorder
+        }
         locationTrackingIdlingResource = RouteProgressStateIdlingResource(
             mapboxNavigation,
             RouteProgressState.TRACKING
@@ -89,7 +104,8 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
                         offRouteLocationUpdate.latitude
                     ),
                     mockRoute.routeWaypoints.last()
-                )
+                ),
+                relaxedExpectedCoordinates = true
             )
         )
         locationTrackingIdlingResource.register()
@@ -103,6 +119,7 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
 
         // start a route
         runOnMainSync {
+            mapboxNavigation.historyRecorder.startRecording()
             mockLocationUpdatesRule.pushLocationUpdate(originLocation)
             mapboxNavigation.startTripSession()
             mapboxNavigation.requestRoutes(
@@ -123,21 +140,23 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
                         reasons: List<RouterFailure>,
                         routeOptions: RouteOptions
                     ) {
-                        // no impl
+                        logE(Tag("DEBUG"), Message("onFailure reasons=$reasons"))
                     }
 
                     override fun onCanceled(
                         routeOptions: RouteOptions,
                         routerOrigin: RouterOrigin
                     ) {
-                        // no impl
+                        logE(Tag("DEBUG"), Message("onCanceled"))
                     }
                 }
             )
         }
 
         // wait for tracking to start
-        Espresso.onIdle()
+        mapboxHistoryTestRule.stopRecordingOnCrash("no location tracking") {
+            Espresso.onIdle()
+        }
         locationTrackingIdlingResource.unregister()
 
         // push off route location and wait for the off route event
@@ -145,13 +164,28 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
         runOnMainSync {
             mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
         }
-        Espresso.onIdle()
+
+        mapboxHistoryTestRule.stopRecordingOnCrash("no off route") {
+            Espresso.onIdle()
+        }
         offRouteIdlingResource.unregister()
 
         // wait for tracking to start again
         locationTrackingIdlingResource.register()
-        Espresso.onIdle()
+
+        mapboxHistoryTestRule.stopRecordingOnCrash("no tracking") {
+            Espresso.onIdle()
+        }
         locationTrackingIdlingResource.unregister()
+
+        runOnMainSync {
+            val countDownLatch = CountDownLatch(1)
+            mapboxNavigation.historyRecorder.stopRecording {
+                logE(Tag("DEBUG"), Message("history path=$it"))
+                countDownLatch.countDown()
+            }
+            countDownLatch.await()
+        }
 
         // assert results
         expectedStates.assert()

--- a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/MapboxHistoryTest.kt
+++ b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/MapboxHistoryTest.kt
@@ -71,16 +71,18 @@ class MapboxHistoryTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.j
     fun setup() {
         Espresso.onIdle()
 
-        mapboxNavigation = MapboxNavigationProvider.create(
-            NavigationOptions.Builder(activity)
-                .accessToken(getMapboxAccessTokenFromResources(activity))
-                .historyRecorderOptions(
-                    HistoryRecorderOptions.Builder()
-                        .build()
-                )
-                .build()
-        )
-        mapboxHistoryTestRule.historyRecorder = mapboxNavigation.historyRecorder
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .historyRecorderOptions(
+                        HistoryRecorderOptions.Builder()
+                            .build()
+                    )
+                    .build()
+            )
+            mapboxHistoryTestRule.historyRecorder = mapboxNavigation.historyRecorder
+        }
         routeCompleteIdlingResource = RouteProgressStateIdlingResource(
             mapboxNavigation,
             RouteProgressState.COMPLETE
@@ -139,7 +141,9 @@ class MapboxHistoryTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.j
         }
 
         // assert and clean up
-        Espresso.onIdle()
+        mapboxHistoryTestRule.stopRecordingOnCrash("no route complete") {
+            Espresso.onIdle()
+        }
         routeCompleteIdlingResource.unregister()
 
         runOnMainSync {

--- a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt
+++ b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteAlternativesTest.kt
@@ -12,6 +12,7 @@ import com.mapbox.navigation.core.MapboxNavigationProvider
 import com.mapbox.navigation.instrumentation_tests.R
 import com.mapbox.navigation.instrumentation_tests.activity.EmptyTestActivity
 import com.mapbox.navigation.instrumentation_tests.utils.MapboxNavigationRule
+import com.mapbox.navigation.instrumentation_tests.utils.history.MapboxHistoryTestRule
 import com.mapbox.navigation.instrumentation_tests.utils.http.MockDirectionsRequestHandler
 import com.mapbox.navigation.instrumentation_tests.utils.idling.FirstLocationIdlingResource
 import com.mapbox.navigation.instrumentation_tests.utils.idling.RouteAlternativesIdlingResource
@@ -31,7 +32,6 @@ import java.util.concurrent.TimeUnit
  * This test ensures that alternative route recommendations
  * are given during active guidance.
  */
-// TODO: Refactor https://github.com/mapbox/mapbox-navigation-android/issues/4429
 class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
 
     @get:Rule
@@ -40,19 +40,25 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
     @get:Rule
     val mockLocationReplayerRule = MockLocationReplayerRule(mockLocationUpdatesRule)
 
+    @get:Rule
+    val mapboxHistoryTestRule = MapboxHistoryTestRule()
+
     private lateinit var mapboxNavigation: MapboxNavigation
 
     @Before
     fun setup() {
-        val routeAlternativesOptions = RouteAlternativesOptions.Builder()
-            .intervalMillis(TimeUnit.SECONDS.toMillis(10))
-            .build()
-        mapboxNavigation = MapboxNavigationProvider.create(
-            NavigationOptions.Builder(activity)
-                .accessToken(getMapboxAccessTokenFromResources(activity))
-                .routeAlternativesOptions(routeAlternativesOptions)
+        runOnMainSync {
+            val routeAlternativesOptions = RouteAlternativesOptions.Builder()
+                .intervalMillis(TimeUnit.SECONDS.toMillis(10))
                 .build()
-        )
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .routeAlternativesOptions(routeAlternativesOptions)
+                    .build()
+            )
+            mapboxHistoryTestRule.historyRecorder = mapboxNavigation.historyRecorder
+        }
     }
 
     @Test
@@ -67,6 +73,7 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
 
         // Start playing locations and wait for an enhanced location.
         runOnMainSync {
+            mapboxNavigation.historyRecorder.startRecording()
             mockLocationReplayerRule.playRoute(routes.first())
             mapboxNavigation.startTripSession()
         }
@@ -81,7 +88,9 @@ class RouteAlternativesTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::cla
         // Wait for route alternatives to be found.
         val alternativesIdlingResource = RouteAlternativesIdlingResource(mapboxNavigation)
         alternativesIdlingResource.register()
-        Espresso.onIdle()
+        mapboxHistoryTestRule.stopRecordingOnCrash("no route alternatives") {
+            Espresso.onIdle()
+        }
 
         // Verify faster alternatives are found.
         val alternatives = alternativesIdlingResource.alternatives!!

--- a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/SanityCoreRouteTest.kt
+++ b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/SanityCoreRouteTest.kt
@@ -15,6 +15,7 @@ import com.mapbox.navigation.core.MapboxNavigationProvider
 import com.mapbox.navigation.instrumentation_tests.activity.EmptyTestActivity
 import com.mapbox.navigation.instrumentation_tests.utils.MapboxNavigationRule
 import com.mapbox.navigation.instrumentation_tests.utils.assertions.RouteProgressStateTransitionAssertion
+import com.mapbox.navigation.instrumentation_tests.utils.history.MapboxHistoryTestRule
 import com.mapbox.navigation.instrumentation_tests.utils.idling.RouteProgressStateIdlingResource
 import com.mapbox.navigation.instrumentation_tests.utils.location.MockLocationReplayerRule
 import com.mapbox.navigation.instrumentation_tests.utils.routes.MockRoutesProvider
@@ -33,6 +34,9 @@ class SanityCoreRouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class
     @get:Rule
     val mockLocationReplayerRule = MockLocationReplayerRule(mockLocationUpdatesRule)
 
+    @get:Rule
+    val mapboxHistoryTestRule = MapboxHistoryTestRule()
+
     private lateinit var mapboxNavigation: MapboxNavigation
 
     private lateinit var routeCompleteIdlingResource: RouteProgressStateIdlingResource
@@ -41,11 +45,14 @@ class SanityCoreRouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class
     fun setup() {
         Espresso.onIdle()
 
-        mapboxNavigation = MapboxNavigationProvider.create(
-            NavigationOptions.Builder(activity)
-                .accessToken(getMapboxAccessTokenFromResources(activity))
-                .build()
-        )
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .build()
+            )
+            mapboxHistoryTestRule.historyRecorder = mapboxNavigation.historyRecorder
+        }
         routeCompleteIdlingResource = RouteProgressStateIdlingResource(
             mapboxNavigation,
             RouteProgressState.COMPLETE
@@ -66,6 +73,7 @@ class SanityCoreRouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class
 
         // execute
         runOnMainSync {
+            mapboxNavigation.historyRecorder.startRecording()
             mockLocationUpdatesRule.pushLocationUpdate {
                 latitude = mockRoute.routeWaypoints.first().latitude()
                 longitude = mockRoute.routeWaypoints.first().longitude()
@@ -106,7 +114,9 @@ class SanityCoreRouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class
         }
 
         // assert and clean up
-        Espresso.onIdle()
+        mapboxHistoryTestRule.stopRecordingOnCrash("no route complete") {
+            Espresso.onIdle()
+        }
         expectedStates.assert()
         routeCompleteIdlingResource.unregister()
     }

--- a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt
+++ b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt
@@ -40,11 +40,13 @@ class AlternativeRouteSelectionTest : BaseTest<BasicNavigationViewActivity>(
 
     @Before
     fun setUp() {
-        mapboxNavigation = MapboxNavigation(
-            NavigationOptions.Builder(activity)
-                .accessToken(getMapboxAccessTokenFromResources(activity))
-                .build()
-        )
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigation(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .build()
+            )
+        }
     }
 
     @After

--- a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/history/MapboxHistoryTestRule.kt
+++ b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/history/MapboxHistoryTestRule.kt
@@ -1,10 +1,15 @@
 package com.mapbox.navigation.instrumentation_tests.utils.history
 
 import android.os.Environment
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
 import com.mapbox.navigation.core.history.MapboxHistoryRecorder
+import com.mapbox.navigation.instrumentation_tests.utils.runOnMainSync
+import com.mapbox.navigation.utils.internal.logE
 import org.junit.rules.TestWatcher
 import org.junit.runner.Description
 import java.io.File
+import java.util.concurrent.CountDownLatch
 
 /**
  * Add this TestRule to your test and the directory will be saved
@@ -46,6 +51,22 @@ class MapboxHistoryTestRule : TestWatcher() {
             val path = description.methodName + File.separator + it.name
             val target = File(directory, path)
             it.copyTo(target)
+        }
+    }
+
+    fun stopRecordingOnCrash(message: String, runner: () -> Unit) {
+        try {
+            runner()
+        } catch (t: Throwable) {
+            runOnMainSync {
+                val countDownLatch = CountDownLatch(1)
+                historyRecorder.stopRecording {
+                    logE(Tag("DEBUG"), Message("$message history path=$it"))
+                    countDownLatch.countDown()
+                }
+                countDownLatch.await()
+            }
+            throw t
         }
     }
 }

--- a/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/http/MockDirectionsRequestHandler.kt
+++ b/instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/utils/http/MockDirectionsRequestHandler.kt
@@ -16,11 +16,16 @@ import okhttp3.mockwebserver.RecordedRequest
 data class MockDirectionsRequestHandler(
     val profile: String,
     val jsonResponse: String,
-    val expectedCoordinates: List<Point>?
+    val expectedCoordinates: List<Point>?,
+    val relaxedExpectedCoordinates: Boolean = false,
 ) : MockRequestHandler {
     override fun handle(request: RecordedRequest): MockResponse? {
-        val prefix =
+        val prefix = if (relaxedExpectedCoordinates) {
+            """/directions/v5/mapbox/$profile"""
+        } else {
             """/directions/v5/mapbox/$profile/${expectedCoordinates.parseCoordinates()}"""
+        }
+
         return if (request.path!!.startsWith(prefix)) {
             MockResponse().setBody(jsonResponse)
         } else {

